### DISCUSSION TITLE
Concurrency?

### DISCUSSION THREAD

**Discussion by prologic** [NicoNex/tau#29] [General]
Do you plan to build this?

If so, how?

What concurrency model fits best with the Tau design thus far?

**Comment by NicoNex** [comment]
Yup I definitely plan to build this and I was thinking of using the Go style, so something like `tau myFunction()`. 
In order for this to work we also need to support channels and have some basic libraries like the Go `sync` one.

**Reply by prologic** [reply]
What if we create a new type called Task?

**Reply by NicoNex** [reply]
What do you have in mind?

**Reply by prologic** [reply]
Sorry I was answering and watching TV at the same time 😅 Here's what I'm thinking...

A few new builtins:

- `pipe()` -- Creates a new channel
- `send(p, any)` -- Send `any` to pipe `p`
- `any = recv(p)` -- Receive `any` from pipe `p`

And a new type (changes to AST):

```
p = pipe()
stop = false
t = task(p, stop) {
  for !stop {
    println(recv(p)
  }
}

send(p, "1")
...
```

Maybe 😅 It's kind of like Go's CSP but Tau-style? 🤔

**Reply by NicoNex** [reply]
I thought about it and I really like the `pipe`, `send` and `recv` builtins.
About `task` I have to think about that, because with a keyword like `tau` you still have normal functions that become coroutines only once you call them with `tau` in front and so it allows to keep the language minimal. 🤔

**Reply by prologic** [reply]
I see I see 🤔 Makes sense

**Comment by NicoNex** [comment]
After a long time I'm glad to announce that the concurrency is finally available on Tau! 😁
https://github.com/NicoNex/tau#concurrency

**Reply by prologic** [reply]
Awesome! 🥳
